kernel/pipes.c

```diff
@@ -405,6 +420,10 @@ int z_impl_k_pipe_put(struct k_pipe *pipe, void *data, size_t bytes_to_write,
 	*bytes_written = pipe_write(pipe, &src_list,
 				    &dest_list, &reschedule_needed);
 
+	if (*bytes_written) {
+		handle_poll_events(pipe);
+	}
+
```

Comment:
Although I think that this covers the most common case, it does break down under one circumstance--all the data written to the pipe went to waiting readers. In such a scenario there is no data for any other thread to receive (since it had all been consumed). This makes me think that the condition(s) to handle poll events should include the pipe's bytes_used field.

Reply:
I had a think about this and changed to:

```
if (pipe->buffer != NULL && pipe->bytes_used) {
	handle_poll_events(pipe);
}
```

If a thread attempts to put bytes in a pipe and there is no space left (ie `*bytes_written` is zero), should the event still trigger? I think it should; after all, there are bytes to be read in the buffer so if anything is polling it should be woken up. In this case, the `*bytes_written` is actually not relevant, and the only things that matters are that a) the pipe has a buffer and b) there are some bytes in the buffer. I will probably add some documentation to state that the waiting on pipes is only possible if the pipe has a ringbuffer behind it, but I am interested to hear your thoughts on this.